File: RAnalytics/Util/Private/_NSString+Encryption.m
#import "_NSString+Encryption.h"
#import <CommonCrypto/CommonDigest.h>

@implementation NSString (Encryption)

-(NSString*) rat_encrypt
{
    NSData *data = [self dataUsingEncoding:NSUTF8StringEncoding];
    NSMutableData *digest = [NSMutableData dataWithLength:CC_SHA256_DIGEST_LENGTH];
    CC_SHA256(data.bytes, (CC_LONG) data.length, digest.mutableBytes);

    NSMutableString *hexDigest = [NSMutableString stringWithCapacity:digest.length * 2];
    const unsigned char *bytes = digest.bytes;

    for (NSUInteger byteIndex = 0; byteIndex < digest.length; ++byteIndex)
    {
       [hexDigest appendFormat:@"%02x", (unsigned int) bytes[byteIndex]];
    }
    return hexDigest;
}

@end

File: RAnalytics/Util/Private/NSString+Encryption.swift
import Foundation
import CommonCrypto.CommonDigest

private extension NSString {
    @objc func ratEncrypt() -> NSString? {
        guard let ratData = nsdata(using: String.Encoding.utf8.rawValue) else { return nil }
        return ratData.digest.hexString
    }

    func nsdata(using encoding: UInt) -> NSData? {
        data(using: encoding) as NSData?
    }
}

private extension NSData {
    var digest: NSData {
        let digestLength = Int(CC_SHA256_DIGEST_LENGTH)
        var hash = [UInt8](repeating: 0, count: digestLength)
        CC_SHA256(bytes, UInt32(length), &hash)
        return NSData(bytes: hash, length: digestLength)
    }

    var hexString: NSString {
        var bytes = [UInt8](repeating: 0, count: length)
        getBytes(&bytes, length: length)
        return bytes.map { String(format: "%02x", $0) }.joined() as NSString
    }
}
